### PR TITLE
Add scaffold for new multiplexer API

### PR DESCRIPTION
**Pull request by Neverlord**
Currently, there's only one implementation for the new multiplexer: `poll`. Other implementations (`epoll`, `kqueue`, etc) will follow once everything else is sorted out.

### CODE REVIEW

File: libcaf_net/src/multiplexer.cpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+size_t multiplexer::num_socket_managers() const noexcept {
+  return managers_.size();
+}
+
+ptrdiff_t multiplexer::index_of(const socket_manager_ptr& mgr) {
+  auto max_index = static_cast<ptrdiff_t>(managers_.size());
+  for (ptrdiff_t index = 0; index < max_index; ++index)
+    if (managers_[index] == mgr)
+      return index;
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/multiplexer.cpp]: It isn't. Thanks for pointing this out.

File: libcaf_net/src/multiplexer.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  if (write_handle_ != invalid_socket) {
+    close(write_handle_);
+    write_handle_ = pipe_socket{};
+  }
+}
+
+bool multiplexer::poll_once(bool blocking) {
+  if (pollset_.empty())
+    return false;
+  // We'll call poll() until poll() succeeds or fails.
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/multiplexer.cpp]: No, `poll` can either succeed at there's at least one event (functions returns `true`) or not (function returns `false`).

File: libcaf_net/src/multiplexer.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        --presult;
+      }
+    }
+    handle_updates();
+    return true;
+  }
+}
+
+void multiplexer::run() {
+  while (!pollset_.empty())
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/multiplexer.cpp]: Exactly.

File: libcaf_net/src/pollset_updater.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  : super(read_handle, std::move(parent)) {
+  mask_ = operation::read;
+}
+
+pollset_updater::~pollset_updater() {
+  // nop
+}
+
+bool pollset_updater::handle_read_event() {
+  for (;;) {
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/pollset_updater.cpp]: Good catch. The `return` belongs into an `else` tree. Then, the function loops until `read` comes back with an error (or `EAGAIN`).

File: libcaf_net/src/socket.cpp [hunk 5]
```diff
@@ -117,12 +180,17 @@ error nonblocking(socket x, bool new_value) {
 
 #endif // CAF_WINDOWS
 
-bool is_error(std::make_signed<size_t>::type res, bool is_nonblock) {
+variant<size_t, sec> check_socket_io_res(std::make_signed<size_t>::type res) {
+  if (res == 0)
```
[the remark this answers is not in the document]
Comment (Neverlord) [reply libcaf_net/src/socket.cpp]: I see that the function name is misleading. The doxygen comment states "`recv` or `send`" (a UDP socket would call `sendto` or `sendmsg`), but the function name simply states `socket_io`.

What if we split into three functions instead?
- `check_stream_socket_io_res`
- `check_pipe_socket_io_res`
- `check_datagram_socket_io_res`

(`check_pipe_socket_io_res` would simply call `check_stream_socket_io_res`, but at least the public API would be cleanly separated)

[the remark this answers is not in the document]
Comment (josephnoir) [reply libcaf_net/src/socket.cpp]: Yeah, I can get behind that.

**Pull request merged**